File: pilhaArray.py
#Pilha com "array" (como Python não tem array, usarei lista)

class No:
    def __init__(self,value):
        self.__value = value

    def get_value(self):
        return self.__value

class Pilha:            
    def __init__(self, lim):
        self.__top = -1
        self.__lim = lim
        self.__array = [None] * self.__lim
        self.__head = self.get_cab()

    #Getters padrões:
    def get_top(self):
        return self.__top
    
    def get_cab(self):
        if self.__array[0] == None:
            return None
        return self.__array[0].get_value()
    
    def get_lenght(self):
        return self.__top + 1
    
    def get_lim(self):
        return self.__lim

    #Setter para dobrar o limite da lista
    def __set_lim(self):
        self.__lim = self.__lim * 2
        newArray = [None] * self.__lim
        for i in range(self.get_lenght()):
            newArray[i] = self.__array[i]
        self.__array = newArray
        return True

    #Retorna os valores contidos na pilha (usei durante o desenvolvimento do programa):
    def get_values(self):
        retorno = ""
        for no in self.__array:
            if no != None:
                retorno = retorno+str(no.get_value())+" "
            else:
                retorno = retorno+str(no)+" "
        return retorno
    
    #Mostra na tela dados gerais da pilha (usei durante o desenvolvimento do programa):
    def dados(self):
        print(f"""
            Topo: {self.get_top()}
            Cab.: {self.get_cab()}
            Len.: {self.get_lenght()}
            Pilha: {self.get_values()}
""")

    #Método de adição de valores à pilha:
    def add(self,value):
        if self.get_lenght() == self.get_lim(): #Se a lista estiver cheia:
            self.__set_lim() #Use o setter para aumentar o limite
        
        no = No(value)
        self.__top = self.__top+1
        self.__array[self.__top] = no
        return True

    #Método de remoção de valores da pilha, que segue o padrão Last-In First-Out:
    def drop(self):
        if self.get_lenght() == 0:
            return False #Pilha vazia
        
        else:
            self.__array[self.__top] = None
            self.__top = self.__top-1
            return True
    
#Função usada para testar o código (como um main()):
def testar():
    pilha = Pilha(5)

    pilha.dados()
    

    for i in range(1,7):
        
        if not pilha.add(i):
            print("PILHA CHEIA!") #Essa mensagem nunca irá aparecer, pois a lista aumenta conforme a demanda
            #print(pilha.get_lenght(), pilha.get_top(), pilha.get_lim())
        pilha.dados()

    for i in range(pilha.get_lenght()+1):
        if not pilha.drop():
            print("PILHA VAZIA!")
        pilha.dados()

testar()

File: pilhaArray.cpp
#include <iostream>
#include <string>

using namespace std;

// Pilha com "array" (como Python não tem array, usarei lista)

class No {
private:
    int value;

public:
    No(int val) {
        value = val;
    }

    int get_value() {
        return value;
    }
};

class Pilha {
private:
    int top;
    int lim;
    No** array;
    int cabecalho;

    // Setter para dobrar o limite da lista
    void set_lim() {
        lim *= 2;
        No** newArray = new No*[lim];
        for (int i = 0; i < get_lenght(); i++) {
            newArray[i] = array[i];
        }
        for (int i = get_lenght(); i < lim; i++) {
            newArray[i] = nullptr;
        }
        delete[] array;
        array = newArray;
    }

public:
    Pilha(int lim) {
        this->lim = lim;
        top = -1;
        array = new No*[lim];
        for (int i = 0; i < lim; i++) {
            array[i] = nullptr;
        }
        cabecalho = get_cab(); // só inicializa, não é usado diretamente
    }

    ~Pilha() {
        for (int i = 0; i <= top; i++) {
            delete array[i];
        }
        delete[] array;
    }

    // Getters padrões:
    int get_top() {
        return top;
    }

    int get_cab() {
        if (array[0] == nullptr) {
            return -1;
        }
        return array[0]->get_value();
    }

    int get_lenght() {
        return top + 1;
    }

    int get_lim() {
        return lim;
    }

    // Retorna os valores contidos na pilha (usei durante o desenvolvimento do programa):
    string get_values() {
        string retorno = "";
        for (int i = 0; i < lim; i++) {
            if (array[i] != nullptr) {
                retorno = retorno+to_string(array[i]->get_value()) + " ";
            } else {
                retorno = retorno+"None ";
            }
        }
        return retorno;
    }

    // Mostra na tela dados gerais da pilha (usei durante o desenvolvimento do programa):
    void dados() {
        cout << "\n"
             << "Topo: " << get_top() << "\n"
             << "Cab.: " << get_cab() << "\n"
             << "Len.: " << get_lenght() << "\n"
             << "Pilha: " << get_values() << "\n"
             << endl;
    }

    // Método de adição de valores à pilha:
    bool add(int value) {
        if (get_lenght() == get_lim()) { // Se a lista estiver cheia:
            set_lim(); // Use o setter para aumentar o limite
        }

        No* no = new No(value);
        top += 1;
        array[top] = no;
        return true;
    }

    // Método de remoção de valores da pilha, que segue o padrão Last-In First-Out:
    bool drop() {
        if (get_lenght() == 0) {
            return false; // Pilha vazia
        } else {
            delete array[top];
            array[top] = nullptr;
            top -= 1;
            return true;
        }
    }
};

// Função usada para testar o código (como um main()):
void testar() {
    Pilha pilha(5);

    pilha.dados();

    for (int i = 1; i <= 6; i++) {
        if (!pilha.add(i)) {
            cout << "PILHA CHEIA!" << endl; // Essa mensagem nunca irá aparecer, pois a lista aumenta conforme a demanda
        }
        pilha.dados();
    }

    for (int i = 0; i <= pilha.get_lenght(); i++) {
        if (!pilha.drop()) {
            cout << "PILHA VAZIA!" << endl;
        }
        pilha.dados();
    }
}

int main() {
    testar();
    return 0;
}
